### PR TITLE
[Form] [Extension] [FormTypeCsrfExtension] Make the csrf field options extensible

### PR DESCRIPTION
**Pull request by esserj**
| Q | A |
| --- | --- |
| Branch? | "master" |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | none |
| License | MIT |
| Doc PR | none |

If you wana decorate the FormTypeCsrfExtension you have no control over options set by the extension, forcing you to reimplement the entire finishView. this way you have a bit more control over it. We required this approach to define custom options declared by another extension

### CODE REVIEW

**Comment by esserj** [comment]
fixed the fail from fabbot.io

**Comment by jakzal** [comment]
Just so I could understand the use case better, what kind of option you needed to customize?

**Comment by esserj** [comment]
We build an extension that requires a by the extension defined option to be set on all fields in order to activate itself on that field. its a client side UI handling submissions through ajax, etc.. we want this process to be opt-in thats why doing it for every formtype by default is not an option

File: src/Symfony/Component/Form/Extension/Csrf/Type/FormTypeCsrfExtension.php [hunk 1]
```diff
@@ -124,6 +124,7 @@ public function configureOptions(OptionsResolver $resolver)
             'csrf_message' => 'The CSRF token is invalid. Please try to resubmit the form.',
             'csrf_token_manager' => $this->defaultTokenManager,
             'csrf_token_id' => null,
+            'csrf_field_options' => array()
```
Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Csrf/Type/FormTypeCsrfExtension.php]: Perhaps define this below `csrf_field_name` for clarification.

File: src/Symfony/Component/Form/Extension/Csrf/Type/FormTypeCsrfExtension.php [hunk 2]
```diff
@@ -105,9 +105,9 @@ public function finishView(FormView $view, FormInterface $form, array $options)
             $tokenId = $options['csrf_token_id'] ?: ($form->getName() ?: get_class($form->getConfig()->getType()->getInnerType()));
             $data = (string) $options['csrf_token_manager']->getToken($tokenId);
 
-            $csrfForm = $factory->createNamed($options['csrf_field_name'], 'Symfony\Component\Form\Extension\Core\Type\HiddenType', $data, array(
+            $csrfForm = $factory->createNamed($options['csrf_field_name'], 'Symfony\Component\Form\Extension\Core\Type\HiddenType', $data, array_replace($options['csrf_field_options'], array(
```
Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Csrf/Type/FormTypeCsrfExtension.php]: Imo. `array_replace_recursive` would be considerable here.

Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Csrf/Type/FormTypeCsrfExtension.php]: Also i'd toggle the arguments, so we can actually override the `mapped` option if needed.

**Comment by ro0NL** [comment]
Perhaps adding `csrf_field_type` makes sense as well.. so we are fully flexible.

**Comment by esserj** [comment]
I've added the suggestions, I see where you wana go with array_replace_recursive, but it seems only usefulll if you by default have a multi-dimentional array and mapped is not, i'd keep that for when you make changes to the default array that require recursive replacements?

File: src/Symfony/Component/Form/Extension/Csrf/Type/FormTypeCsrfExtension.php [hunk 3]
```diff
@@ -121,9 +121,11 @@ public function configureOptions(OptionsResolver $resolver)
         $resolver->setDefaults(array(
             'csrf_protection' => $this->defaultEnabled,
             'csrf_field_name' => $this->defaultFieldName,
+            'csrf_field_type' => 'Symfony\Component\Form\Extension\Core\Type\HiddenType',
+            'csrf_field_options' => array(),
             'csrf_message' => 'The CSRF token is invalid. Please try to resubmit the form.',
             'csrf_token_manager' => $this->defaultTokenManager,
-            'csrf_token_id' => null,
+            'csrf_token_id' => null
```
Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Csrf/Type/FormTypeCsrfExtension.php]: Last comma is required code style, you should revert this line :)

**Comment by esserj** [comment]
fixed that, must take some time to go through the code style, only jumped on the symfony wagon recently :)

**Comment by patrick-mcdougle** [comment]
@esserj 

>we want this process to be opt-in thats why doing it for every formtype by default is not an option

Is this different than disabling it?

**Comment by esserj** [comment]
@patrick-mcdougle an opt-out strategy would in our case have the same issue

**Comment by patrick-mcdougle** [comment]
I mean, how is this different than disabling csrf on some of your forms? Or am I not understanding the use case?

**Comment by esserj** [comment]
Why would we want to disable CSRF? This is a CMS back-end where admins will be working in. If there is one thing we want in there its CSRF. so we would have to roll my own version or decorate the current implementation (which we did now) but any future changes/improvements will have to manually be ported. And thats when I decided to create the PR as it will address the non-extensible nature of the current implementation

**Comment by patrick-mcdougle** [comment]
I'm just not clear on what you're trying to do. I didn't follow your explanation to @jakzal Why do you need the csrf token field to be rendered using something different than a hidden field?

**Comment by esserj** [comment]
@patrick-mcdougle I don't need anything different than a hidden field, that was a suggestion from @ro0NL that i took up in this extensibility patch, i need to be able to add an option to the csrf field

File: src/Symfony/Component/Form/Extension/Csrf/Type/FormTypeCsrfExtension.php [hunk 4]
```diff
@@ -105,9 +105,9 @@ public function finishView(FormView $view, FormInterface $form, array $options)
             $tokenId = $options['csrf_token_id'] ?: ($form->getName() ?: get_class($form->getConfig()->getType()->getInnerType()));
             $data = (string) $options['csrf_token_manager']->getToken($tokenId);
 
-            $csrfForm = $factory->createNamed($options['csrf_field_name'], 'Symfony\Component\Form\Extension\Core\Type\HiddenType', $data, array(
+            $csrfForm = $factory->createNamed($options['csrf_field_name'], $options['csrf_field_type'], $data, array_replace(array(
```
Comment (xabbuh) [inline src/Symfony/Component/Form/Extension/Csrf/Type/FormTypeCsrfExtension.php]: While I can understand that there may be use case where you need to (want to) customise the CSRF field's options, I do not see a need to change the type being used. Without a really good use case I am against making it configurable.

Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Csrf/Type/FormTypeCsrfExtension.php]: Simple, allow for `MyHiddenType::class`. Imo. it's weird to allow passing options, but not allowing a custom type with _default_ options.

Comment (xabbuh) [inline src/Symfony/Component/Form/Extension/Csrf/Type/FormTypeCsrfExtension.php]: There shouldn't be a custom hidden type. If you are creating a custom hidden type, you will very likely better be off with a form type extension.

Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Csrf/Type/FormTypeCsrfExtension.php]: Maybe your're right.. i have no actual case for this. Then again, forcing a specific type doesnt feel extensible for an _extension_...

Comment (fabpot) [inline src/Symfony/Component/Form/Extension/Csrf/Type/FormTypeCsrfExtension.php]: I agree with @xabbuh

**Comment by HeahDude** [comment]
Hi @esserj,  thank you for proposing this change, but the use case it supported can already be solved.

>If you wana decorate the FormTypeCsrfExtension you have no control over options set by the extension,

Since 3.2 it is possible to handle priorities on form type extension when registering them as services, this feature has been added so it can be easier to rely on other extensions definitions.

Hence I don't think we need to add yet another option to handle this, you could easily override options from another type extension now. Using the Form component stand alone should already provide full control on the registration order of different type extensions.

If there is no other concrete use cases, I'm 👎 for now.

**Comment by fabpot** [comment]
Closing this PR as there is no more activity and several concerns were raised.

**Pull request closed without merging**